compiler/src/iree/compiler/Dialect/LinalgExt/IR/AggregatedOpInterfaceImpl.cpp

```diff
@@ -890,24 +921,49 @@ FailureOr<SmallVector<Value>> Im2colOp::decomposeOperation(OpBuilder &b) {
   sliceOffsets = SmallVector<OpFoldResult>(getOutputRank(), zero);
   sliceSizes = SmallVector<OpFoldResult>(getOutputRank(), one);
   sliceStrides = SmallVector<OpFoldResult>(getOutputRank(), one);
-  sliceSizes.back() = kTileSize;
-  for (auto [idx, iv] : llvm::enumerate(ivs)) {
-    sliceOffsets[idx] = iv;
-  }
-  // Insert a `linalg.copy` so there is something to vectorize in the
-  // decomposition. Without this copy, the extract and insert slice ops
-  // do not get vectorized, and the sequence becomes a scalar memref.copy.
-  // This memref.copy could be vectorized after bufferization, but it is
-  // probably better to vectorize during generic vectorization.
-  Value copyDest = b.create<tensor::ExtractSliceOp>(
-      nestedLoc, extractType, loopNest.loops.back().getRegionIterArg(0),
+
+  Value inputForInsert;
+  // If the im2col input tensor has the batch dim at last and can be vectorized,
+  // transpose the dimensions to match the output order (Batch, M, K).
+  if (vectorizeInnerBLoop) {
+    SmallVector<int64_t> transposePerm;
+    transposePerm.insert(transposePerm.end(), getBatchPos().begin(),
+                         getBatchPos().end());
+    transposePerm.insert(transposePerm.end(), getMPos().begin(),
+                         getMPos().end());
+    transposePerm.insert(transposePerm.end(), getKPos().begin(),
+                         getKPos().end());
```

Comment:
nit: use `append` which improves readability. IMO, it reduces one level of logic.